Document UTxO-HD and cleanup (#410)
input-output-hk/ouroboros-consensus: CONTRIBUTING.md

Author: I removed the makefile in favor of improving the haddocks script. Good point.

Author: The makefile is gone so this should be easier now

Note the second one cannot be used when we want to provide CLI arguments to the
test-suite.

> # Generating documentation and setting up hoogle

Reviewer: We should create an issue for providing a `nix` alternative to these steps (and assign it to @amesgen :grimacing: )

Documentation contains some tikz figures that require some preprocessing for

> them to be displayed. In order to do this, use the documentation script:

Reviewer: ```suggestion
them to be displayed. To do this, use the documentation script:
```

I tend to prefer the equivalent "to" to keep the sentences concise.

```bash
./scripts/docs/haddocks.sh
```

This will install `latex-svg-image` from [a
fork](https://github.com/jasagredo/latex-svg/tree/jasagredo/draw-tikz/latex-svg-image),
and `cabal-docspec` from a binary (do inspect the `Makefile` to see what it
does), then building the haddocks for your package.

Often times, it is useful to have a `hoogle` server at hand, with the packages
and its dependencies. For this, one needs to first modify the global cabal
configuration:

```bash

> cabal user-config update -a "documentation: True

Reviewer: Isn't it better to set this per-project in a `cabal.project.local`?

Author: With latest versions this can be omitted.

Author: Although in any case, in previous versions you need this to have the documentation of the dependencies.

haddock
  html: True

>   hoogle: True"

Reviewer: Is this the output of the bash command? If so we might want to distinguish it from the commands.

Author: it is not. That command is literally as one has to write it. It contains a multiline string.

Note the second one cannot be used when we want to provide CLI arguments to the
test-suite.

# Generating documentation and setting up hoogle

Documentation contains some tikz figures that require some preprocessing for

> them to be displayed. In order to do this, one can just invoke the `make` rules:

Reviewer: ```suggestion
them to be displayed. To do this, one can just invoke the `make` rules:
```

> ```bash

Reviewer: I wonder if this `Makefile` should go inside the `docs` folder. People might get confused and think they can build/test the code using it.

Author: The makefile is gone now

make haddocks

# optionally provide a specific package or component
make haddocks pkg=ouroboros-consensus
```

> This will install `latex-svg-image` from [a

Reviewer: We should probably ask @amesgen for help `nix`ifying this.

fork](https://github.com/jasagredo/latex-svg/tree/jasagredo/draw-tikz/latex-svg-image),

> and `cabal-docspec` from a binary (do inspect the `Makefile` to see what it
> does), then building the haddocks for your package.

Reviewer: ```suggestion
and `cabal-docspec` from a binary and then build the haddocks for your package.
```

Often times, it is useful to have a `hoogle` server at hand, with the packages
and its dependencies. For this, one needs to first modify the global cabal
configuration:

```bash
cabal user-config update -a "documentation: True
haddock
  html: True
  hoogle: True"
```

and then invoke the `make` rules again:

```bash
make hoogle
```

This will fire a `hoogle` server at https://localhost:8080/ with the local

> packages and their dependencies.

Author: Hoogle cannot be deployed on github pages :cry:
